### ISSUE TITLE
[Bug] [zos_data_set] Not able to create data sets with names that include permitted symbols

### ISSUE THREAD

**Issue by ddimatos** [ansible-collections/ibm_zos_core#642]
### Bug description

User tries to create a data set in Ansible using this name "JYVESB.A$A.DBRM" , review the naming conventions [here.](https://www.ibm.com/docs/en/zos/2.1.0?topic=sets-data-set-naming-rules) what gets created is "'JYVESB.A.DBRM'"

After some added further analysis we found that the dependency APIs need escaping for special chars when our Ansible code detects a `$, #, @` character in order to create data sets with special characters. 

The Ansible better arg parser will need to infer that a special character has been entered continue as it does today which is to validate the input but if it see a special character it should track that and then escape it for the user. As it is, the user can not escape it because it would violate our permitted special characters that is why our code should handle it. 

It could be more involved than just the better arg parser as some modules are handling validation logic and arg passing directly, this could be a fairly sizable effort to ensure all the code is handling this correctly.

I have included two examples , one using ds type LIBRARY and one PDSE which are the same thing.

Recreate below:
```
- hosts: zvm
  gather_facts: false
  environment: "{{ environment_vars }}"
  collections:
    - ibm.ibm_zos_core

  tasks:
    - name: "Create data set with $ failed: 'JYVESB.A$A.DBRM' -> Created PDS JYVESB.A.DBRM"
      zos_data_set:
        name: 'JYVESB.A$A.DBRM'
        type: LIBRARY
        space_primary: 10
        space_secondary: 20
        space_type: CYL
        record_format: U
        record_length: 0
      ignore_errors: true
```

Python recreate:
Python code recreate:
1) As a LIBRARY:
```
from zoautil_py import datasets
from zoautil_py.types import Dataset
from pprint import pprint
name="JYVESB.A$A.DBRM"
type="LIBRARY"
kwargs = {'primary_space':'10CYL', 'secondary_space':'20CYL', 'record_format':'U', 'record_length':'0'}
response = datasets.create(name, type, **kwargs)
pprint(vars(response))
{'block_size': '0',
 'dsorg': 'PO',
 'last_referenced': '0000/01/01',
 'lrecl': '0',
 'name': 'JYVESB.A.DBRM',
 'recfm': 'U',
 'total_space': '8499600',
 'used_space': '??',
 'volume': '000000',
 'vrbs': 0} 
```
2) As a PDSE , building on top of shell code from (1)
```
type="PDSE"
response = datasets.create(name, type, **kwargs)
pprint(vars(response))
{'block_size': '0',
 'dsorg': 'PO',
 'last_referenced': '0000/01/01',
 'lrecl': '0',
 'name': 'JYVESB.A.DBRM',
 'recfm': 'U',
 'total_space': '8499600',
 'used_space': '??',
 'volume': 'SCR03',
 'vrbs': 0} 
```

UPDATE: Python code that corrects this behavior:
Case 1:
```
>>> from zoautil_py import datasets
>>> from zoautil_py.types import Dataset
>>> from pprint import pprint
>>> type="LIBRARY"
>>> kwargs = {'primary_space':'10CYL', 'secondary_space':'20CYL', 'record_format':'U', 'record_length':'0'}
>>> name="JYVESB.\$\$TEST"
>>> response = datasets.create(name, type, **kwargs)
>>> pprint(vars(response))
{'block_size': '0',
 'dsorg': 'PO',
 'last_referenced': '0000/01/01',
 'lrecl': '0',
 'name': 'JYVESB.$$TEST',
 'recfm': 'U',
 'total_space': '8499600',
 'used_space': '??',
 'volume': '333333',
 'vrbs': 0}
```

Case 2:
```
>>> from zoautil_py import datasets
>>> from zoautil_py.types import Dataset
>>> from pprint import pprint
>>> type="LIBRARY"
>>> kwargs = {'primary_space':'10CYL', 'secondary_space':'20CYL', 'record_format':'U', 'record_length':'0'}
>>> name="JYVESB.A\$A.DBRM"
>>> response = datasets.create(name, type, **kwargs)
>>> pprint(vars(response))
{'block_size': '0',
 'dsorg': 'PO',
 'last_referenced': '0000/01/01',
 'lrecl': '0',
 'name': 'JYVESB.A$A.DBRM',
 'recfm': 'U',
 'total_space': '8499600',
 'used_space': '??',
 'volume': '000000',
 'vrbs': 0} 
```

### Playbook verbosity output

```
TASK [Create data set with $ failed: 'JYVESB.A$A.DBRM' -> Created PDS JYVESB.A.DBRM] *************************************************************************************************************
changed: [zvm]
```
No output because we ignored the error. I hacked in a strategic dump to see what was sent to the dependency you can in module utils that looks like this (see commented out lines) that helped me reproduce:
```
        original_args = locals()
        formatted_args = DataSet._build_zoau_args(**original_args)
       # kwargsString =  ", ".join(f"{key}={value}" for key, value in formatted_args.items())
       # raise Exception(kwargsString)
        response = datasets._create(**formatted_args)
        if response.rc > 0:
            raise DatasetCreateError(
                name, response.rc, response.stdout_response + response.stderr_response
            )
        return response.rc
```
Hack returns:
```
TASK [Create data set with $ failed: 'JYVESB.A$A.DBRM' -> Created PDS JYVESB.A.DBRM] *************************************************************************************************************
fatal: [zvm]: FAILED! => {"changed": false, "message": "", "msg": "Exception('name=JYVESB.A$A.DBRM, type=LIBRARY, primary_space=10CYL, secondary_space=20CYL, record_format=U, record_length=0')", "names": ["JYVESB.A$A.DBRM"]}
```

### Ansible version

```shell
2.11.12
```

### IBM z/OS Ansible core Version

v1.4.0

### IBM ZOAU version

v1.2.2

### z/OS version

2.5

### Ansible module

zos_data_set

**Comment by vlodeibm** [comment]
@ddimatos hi, just to precise that all type of datasets are concerned, I mean also sequential (type:SEQ) files and  VSAM data set :(

**Comment by richp405** [comment]
observation:
dtouch is a good example of this, and it doesn't quite match the manual:
bash-4.3# dtouch JOHN.POUND\$.PDS
bash-4.3# dls JOHN.PO*
JOHN.POUND$.PDS

bash-4.3# dtouch JYVESB.A$A.DBRM
bash-4.3# dls JYVESB.*
JYVESB.A.DBRM

bash-4.3# dtouch JYVESB.A\$A.DBRM
bash-4.3# dls JYVESB.*
JYVESB.A$A.DBRM
JYVESB.A.DBRM

but pound symbol chokes on escape, and works un-escaped?

bash-4.3# dtouch "JOHN.POUND#.PDS"
bash-4.3# dls "JOHN.PO*"
JOHN.POUND#.PDS

bash-4.3# dtouch "JOHN.POUND\#.PDS"
BGYSC1402E Dynalloc Failure. Error Code: 0x35c, Information Code: 0x02

@ symbol works either way - escaped or not

dls JYVESB.*
JYVESB.A@A.DBRM
bash-4.3# drm JYVESB.A\@A.DBRM
bash-4.3# dls JYVESB.*
BGYSC1103E No datasets match pattern: JYVESB.*.

So maybe it's "ignore #-passthrough, and escape on $ and @"?

I've added the test code to job.py, and will call it a few places for base testing.:

**Comment by richp405** [comment]
I have a fix that works for ds creation

It does NOT work for deletion, because the implementation is slightly different:
creation uses dtouch + options + cleaneddsname (works fine)
deletion uses drm + options + "cleaneddsname" (trips over some escapes)

Not sure how to work around this yet.

**Comment by richp405** [comment]
We have validated a code change that will properly escape the characters, but need to wait for the matching component in zoau to handle this change.  Ticket filed is Jira/Narare/10103

**Comment by ddimatos** [comment]
This relies on ZOAU 1.3 , IBM z/OS core collection will support 1.3 in Q1 2024

**Comment by richp405** [comment]
Going to put the symbol-escape login into PR for Friday... it won't do any harm to have it ready, even if we ice it until closer to 1.3 release.

**Comment by ddimatos** [comment]
Based on the discussion we had today, we decided this work needs careful consideration and should not be included in 1.8.0-beta.1 and instead try for 1.9.0-beta.1 which will still be a NOOP until zoua 1.3 but the code code be ready for early build testing. Moving this to Q4.5 since there is capacity for you there and will post plan comment this work as it is out of plan.

**Comment by fernandofloresg** [comment]
This is completed by #1371
